vrf: Add support of Virtual Routing and Forwarding

cathay4t opened the pull request:
Example of desire state for VRF:

```yml
interfaces:
- name: vrf0
  type: vrf
  vrf:
    route-table-id: 100
    port:
      - eth1
      - eth2
```

Example yaml files(`vrf0_with_eth1.yml` and `vrf0_absent.yml`) included.

Due to limitation of NetworkManager, `NmstateNotSupportedError` will be
raised when changing route table ID of existing interface.

Test cases included.

EdDev commented on libnmstate/schema.py:
```diff
@@ -348,3 +349,8 @@ class Neighbors:
+
+class VRF:
+    CONFIG_SUBTREE = "vrf"
+    PORT_SUBTREE = "ports"
```
We have currently a mixture of "port" (bridge) and "ports" (team), perhaps we should sync them all to be consistent.

I do not really know which is better, plural or singular, there are some conventions floating around. Maybe just check with some other projects that use xml/yaml/json for the configuration. Ansible network unified ones?

cathay4t replied:
@ffmancera Can you also include this into your `inclusive language plan`?

cathay4t replied:
I have changed to `port` to align with LinuxBridge and OvsBridge.

Since there might no nmstate consumer using team, so we will deprecate the use of `ports` in team: https://github.com/nmstate/nmstate/issues/1328

EdDev reviewed (commented):
The scheme looks good to me. One small thing to consider I commented inline.
Thanks!

ffmancera reviewed (approved):
LGTM, thanks!

ffmancera commented:
@smyle10 Please, verify this when possible. Thanks!

smyle10 commented:
I tried with 
nmstate-0.4.1-0.20200918141918765851.pr1310.17.g460ea50.el8.noarch
NetworkManager-1.26.0-7.el8.x86_64
DISTRO=RHEL-8.3.0-20200918.n.1
```
interfaces:
- name: vrf0
  type: vrf
  state: up
  vrf:
    route-table-id: 100
    port:
      - veth1
      - veth0
```
It got a problem:
```
# nmstatectl set vrf.yaml 
thread '<unnamed>' panicked at 'not yet implemented: netlink-rust does not support parsing this yet', src/lib/route.rs:441:17
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace.
fatal runtime error: failed to initiate panic, error 5
Aborted (core dumped)
```

cathay4t commented:
> I tried with
> nmstate-0.4.1-0.20200918141918765851.pr1310.17.g460ea50.el8.noarch
> NetworkManager-1.26.0-7.el8.x86_64
> DISTRO=RHEL-8.3.0-20200918.n.1
> 
> ```
> interfaces:
> - name: vrf0
>   type: vrf
>   state: up
>   vrf:
>     route-table-id: 100
>     port:
>       - veth1
>       - veth0
> ```
> 
> It got a problem:
> 
> ```
> # nmstatectl set vrf.yaml 
> thread '<unnamed>' panicked at 'not yet implemented: netlink-rust does not support parsing this yet', src/lib/route.rs:441:17
> note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace.
> fatal runtime error: failed to initiate panic, error 5
> Aborted (core dumped)
> ```

The problem has been solved in nispor package and not related to nmstate VRF feature.
Can you try it on other servers?

Thank you.

cathay4t committed:
vrf: Add support of Virtual Routing and Forwarding

Example of desire state for VRF:

```yml
interfaces:
- name: vrf0
  type: vrf
  vrf:
    route-table-id: 100
    port:
      - eth1
      - eth2
```

Example yaml files(`vrf0_with_eth1.yml` and `vrf0_absent.yml`) included.

Due to limitation of NetworkManager, `NmstateNotSupportedError` will be
raised when changing route table ID of existing interface.

Test cases included.

Signed-off-by: Gris Ge <fge@redhat.com>

smyle10 commented:
Tested basic operations
- Create, remove the main interface
- Add/remove subordinates
- IP address

cathay4t commented:
Timeout on waiting consumer QE. Will create new PR if they found any problem.
